src/linearAlgebra/BlasLapackKernels.cpp

```diff
@@ -9,59 +9,75 @@ namespace dftefe
           }
       }
 
 
-      template <typename ValueType, dftefe::utils::MemorySpace memorySpace>
+      template <dftefe::utils::MemorySpace memorySpace,
+                typename ValueType1,
+                typename ValueType2>
       void
-      Kernels<ValueType, memorySpace>::hadamardProduct(const size_type  size,
-                                                       const ValueType *x,
-                                                       const ValueType *y,
-                                                       ValueType *      z)
+      Kernels<memorySpace, ValueType1, ValueType2>::hadamardProduct(
+        const size_type                      size,
+        const ValueType1 *                   x,
+        const ValueType2 *                   y,
+        scalar_type<ValueType1, ValueType2> *z)
       {
         for (size_type i = 0; i < size; ++i)
           {
-            z[i] = x[i] * y[i];
+            z[i] = ((scalar_type<ValueType1, ValueType2>)x[i]) *
+                   ((scalar_type<ValueType1, ValueType2>)y[i]);
           }
       }
 
-      template <typename ValueType, dftefe::utils::MemorySpace memorySpace>
+      template <dftefe::utils::MemorySpace memorySpace,
+                typename ValueType1,
+                typename ValueType2>
       void
-      Kernels<ValueType, memorySpace>::axpby(const size_type  size,
-                                             const ValueType  alpha,
-                                             const ValueType *x,
-                                             const ValueType  beta,
-                                             const ValueType *y,
-                                             ValueType *      z)
+      Kernels<memorySpace, ValueType1, ValueType2>::axpby(
+        const size_type                           size,
+        const scalar_type<ValueType1, ValueType2> alpha,
+        const ValueType1 *                        x,
+        const scalar_type<ValueType1, ValueType2> beta,
+        const ValueType2 *                        y,
+        scalar_type<ValueType1, ValueType2> *     z)
       {
         for (size_type i = 0; i < size; ++i)
           {
-            z[i] = alpha * x[i] + beta * y[i];
+            z[i] = ((scalar_type<ValueType1, ValueType2>)alpha) *
+                     ((scalar_type<ValueType1, ValueType2>)x[i]) +
+                   ((scalar_type<ValueType1, ValueType2>)beta) *
+                     ((scalar_type<ValueType1, ValueType2>)y[i]);
           }
       }
 
 
-      template <typename ValueType, dftefe::utils::MemorySpace memorySpace>
+      template <dftefe::utils::MemorySpace memorySpace,
+                typename ValueType1,
+                typename ValueType2>
       std::vector<double>
-      Kernels<ValueType, memorySpace>::amaxsMultiVector(
-        const size_type  vecSize,
-        const size_type  numVec,
-        const ValueType *multiVecData)
+      Kernels<memorySpace, ValueType1, ValueType2>::amaxsMultiVector(
```

Comment:
Since this one of the member functions that requires only ValueType1, I think it should just be `template <utils::MemorySpace memorySpace, typename ValueType1> Kernels<memorySpace, ValueType1>::amaxsMultiVector()`. Adding `ValueType2` might lead to compilation error if the user side is calling it as `Kernels<memorySpace, ValueType1>::amaxsMultiVector()` (i.e., taking `ValueType2` to be same as `ValueType1`). The compiler won't find the definition for  `Kernels<memorySpace, ValueType1>::amaxsMultiVector()`

Reply:
Using
`template <utils::MemorySpace memorySpace, typename ValueType1> Kernels<memorySpace, ValueType1>::amaxsMultiVector()`

doesn't compile with the error "invalid use of incomplete type ‘class dftefe::linearAlgebra::blasLapack::Kernels<memorySpace, ValueType1>’"

> The compiler won't find the definition for Kernels<memorySpace, ValueType1>::amaxsMultiVector()
The compiler already finds its as I am explicitly instantiating all the possible classes from the template class  in the .cpp and .cu files.

Reply:
Okay. I was mistaken. I am not too bothered by the duplicate compilation, as it only increases the compilation time. I am okay with the previous Kernels and KernelTwoValueTypes approach as well. But I am afraid that if there is a need for a third Valuetype in future, it will just make things difficult to maintain.

Reply:
The issue with the current approach when adding a third Valuetype is that all existing kernel definitions for one and two value types needs to be touched (the way we write the template for them in .cpp and .cu). In that respect the other PR is more cleaner as we only have to write KernelsThreeValueTypes class while leaving the previous classes untouched.